File: package-upgrade/scripts/python/ast_scanner.py
#!/usr/bin/env python3
"""AST scanner for package usage analysis.

Usage: python ast_scanner.py <project_path> <package_name>

Output JSON (aligned with ast_scanner_js.js / ast_scanner_go.go):
  {
    "scan_results": [...],        // [] when no matches — never null
    "total_files": N,             // .py files walked
    "files_scanned": N,           // alias of total_files for clarity
    "import_count": N,            // total import sites across all files
    "usage_count": N,             // total usage symbols across all files
    "package_name": "...",
    "language": "python",
    "warnings": [...],            // [] when none
    "verdict": "zero_impact" | "has_impact" | "scan_errored",
    "verdict_reason": "..."
  }

The `verdict` field lets the skill (Phase 4) short-circuit without
re-deriving conclusions from the raw array. Matches the schema first
introduced in ast_scanner_go.go.
"""

import ast
import json
import sys
from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple


class PackageUsageVisitor(ast.NodeVisitor):
    """AST visitor to track package imports and usage."""

    def __init__(self, package_name: str, source_lines: List[str]):
        self.package_name = package_name
        self.source_lines = source_lines
        self.imports = []  # Import statements
        self.usages = []  # Symbol usage locations
        self.imported_names = {}  # Maps local name -> original module path

    def visit_Import(self, node: ast.Import):
        """Track 'import module' statements."""
        for alias in node.names:
            if alias.name == self.package_name or alias.name.startswith(f"{self.package_name}."):
                local_name = alias.asname or alias.name
                self.imported_names[local_name] = alias.name
                self.imports.append(
                    {
                        "type": "import",
                        "module": alias.name,
                        "alias": alias.asname,
                        "line": node.lineno,
                        "context": self._get_context(node.lineno),
                    }
                )
        self.generic_visit(node)

    def visit_ImportFrom(self, node: ast.ImportFrom):
        """Track 'from module import name' statements."""
        if node.module and (
            node.module == self.package_name or node.module.startswith(f"{self.package_name}.")
        ):
            for alias in node.names:
                local_name = alias.asname or alias.name
                full_name = f"{node.module}.{alias.name}" if alias.name != "*" else node.module
                self.imported_names[local_name] = full_name
                self.imports.append(
                    {
                        "type": "from_import",
                        "module": node.module,
                        "name": alias.name,
                        "alias": alias.asname,
                        "line": node.lineno,
                        "context": self._get_context(node.lineno),
                    }
                )
        self.generic_visit(node)

    def visit_Attribute(self, node: ast.Attribute):
        """Track attribute access like obj.method()."""
        chain = self._resolve_attr_chain(node)
        if chain:
            root = chain.split(".")[0]
            if root in self.imported_names:
                full_name = self.imported_names[root] + chain[len(root) :]
                self.usages.append(
                    {
                        "symbol": full_name,
                        "line": node.lineno,
                        "context": self._get_context(node.lineno),
                    }
                )
        self.generic_visit(node)

    def visit_Name(self, node: ast.Name):
        """Track name references."""
        if node.id in self.imported_names:
            self.usages.append(
                {
                    "symbol": self.imported_names[node.id],
                    "line": node.lineno,
                    "context": self._get_context(node.lineno),
                }
            )
        self.generic_visit(node)

    def _resolve_attr_chain(self, node: ast.AST) -> Optional[str]:
        """Resolve attribute chain like obj.method.attr to string."""
        parts = []
        while isinstance(node, ast.Attribute):
            parts.append(node.attr)
            node = node.value
        if isinstance(node, ast.Name):
            parts.append(node.id)
            return ".".join(reversed(parts))
        return None

    def _get_context(self, lineno: int, radius: int = 5) -> str:
        """Get code context around a line number (±radius lines)."""
        start = max(0, lineno - radius - 1)
        end = min(len(self.source_lines), lineno + radius)
        lines = self.source_lines[start:end]
        return "\n".join(f"{start + i + 1:4d} | {line}" for i, line in enumerate(lines))


def _scan_file_internal(filepath: Path, package_name: str) -> Tuple[Optional[Dict[str, Any]], bool]:
    """Scan a single Python file for package usage.

    Returns (result_or_None, parse_failed). `parse_failed` is True when the
    file was unreadable or unparseable; the caller needs this to distinguish
    `zero_impact` (clean walk, no match) from `scan_errored` (couldn't parse).
    """
    try:
        source = filepath.read_text(encoding="utf-8")
        tree = ast.parse(source, filename=str(filepath))
    except (SyntaxError, UnicodeDecodeError):
        return None, True

    visitor = PackageUsageVisitor(package_name, source.splitlines())
    visitor.visit(tree)

    if not visitor.imports and not visitor.usages:
        return None, False

    return {
        "file": str(filepath),
        "imports": visitor.imports,
        "usages": visitor.usages,
    }, False


def scan_file(filepath: Path, package_name: str) -> Optional[Dict[str, Any]]:
    """Scan a single Python file for package usage.

    Backwards-compatible wrapper that hides the parse-failed flag.
    """
    result, _ = _scan_file_internal(filepath, package_name)
    return result


# Skip these directories during project walk.
_SKIP_DIRS = {".git", "__pycache__", ".venv", "venv", "node_modules", ".tox", ".pytest_cache"}


def _walk_python_files(project_path: str):
    """Yield .py files under project_path, skipping ignored directories."""
    project_root = Path(project_path)
    for py_file in project_root.rglob("*.py"):
        if any(part in _SKIP_DIRS or part.startswith(".") for part in py_file.parts):
            continue
        yield py_file


def scan_project_stats(
    project_path: str, package_name: str
) -> Tuple[List[Dict[str, Any]], int, int]:
    """Walk project, return (matching_results, files_scanned, parse_errors)."""
    results: List[Dict[str, Any]] = []
    files_scanned = 0
    parse_errors = 0
    for py_file in _walk_python_files(project_path):
        files_scanned += 1
        result, parse_failed = _scan_file_internal(py_file, package_name)
        if parse_failed:
            parse_errors += 1
        elif result:
            results.append(result)
    return results, files_scanned, parse_errors


def scan_project(project_path: str, package_name: str) -> List[Dict[str, Any]]:
    """Scan entire project for package usage (results only — back-compat)."""
    results, _, _ = scan_project_stats(project_path, package_name)
    return results


def compute_verdict(
    results: List[Dict[str, Any]],
    files_scanned: int,
    parse_errors: int,
    package_name: str,
) -> Tuple[str, str, int, int]:
    """Mirror ast_scanner_go.go verdict logic.

    Returns (verdict, verdict_reason, import_count, usage_count).
    """
    import_count = sum(len(r["imports"]) for r in results)
    usage_count = sum(len(r["usages"]) for r in results)

    if import_count == 0 and usage_count == 0 and parse_errors == 0:
        return (
            "zero_impact",
            f"scanned {files_scanned} .py file(s); no import or usage of " f"{package_name} found",
            import_count,
            usage_count,
        )
    if import_count == 0 and usage_count == 0 and parse_errors > 0:
        return (
            "scan_errored",
            f"{parse_errors} file(s) failed to parse; no matches in the "
            f"remaining {files_scanned - parse_errors}. Treat as inconclusive.",
            import_count,
            usage_count,
        )
    return (
        "has_impact",
        f"{import_count} import site(s) and {usage_count} usage(s) found "
        f"across {len(results)} file(s)",
        import_count,
        usage_count,
    )


def main():
    if len(sys.argv) < 3:
        print("Usage: python ast_scanner.py <project_path> <package_name>", file=sys.stderr)
        sys.exit(1)

    project_path = sys.argv[1]
    package_name = sys.argv[2]

    results, files_scanned, parse_errors = scan_project_stats(project_path, package_name)
    verdict, verdict_reason, import_count, usage_count = compute_verdict(
        results, files_scanned, parse_errors, package_name
    )

    warnings: List[str] = []
    if parse_errors > 0:
        warnings.append(f"{parse_errors} file(s) failed to parse; results may be incomplete")

    output = {
        "scan_results": results,
        "total_files": files_scanned,
        "files_scanned": files_scanned,
        "import_count": import_count,
        "usage_count": usage_count,
        "package_name": package_name,
        "language": "python",
        "warnings": warnings,
        "verdict": verdict,
        "verdict_reason": verdict_reason,
    }

    print(json.dumps(output, indent=2))


if __name__ == "__main__":
    main()

File: package-upgrade/scripts/javascript/ast_scanner.js
#!/usr/bin/env node
/**
 * ast_scanner_js.js — Scan JS/TS project for usage of a target package.
 *
 * Usage:
 *   node ast_scanner_js.js <project_path> <package_name>
 *
 * Output JSON (aligned with ast_scanner.py / ast_scanner_go.go):
 *   {
 *     "scan_results": [
 *       {
 *         "file": "src/api/client.ts",
 *         "imports": [
 *           { "type": "esm_default", "module": "axios", "alias": "axios",
 *             "line": 3, "context": "..." }, ...
 *         ],
 *         "usages": [
 *           { "symbol": "axios.get", "line": 12, "context": "..." }, ...
 *         ]
 *       }, ...
 *     ],
 *     "total_files":   N,        // .js/.ts files walked
 *     "files_scanned": N,        // alias of total_files
 *     "import_count":  N,        // total import sites across all files
 *     "usage_count":   N,        // total usage symbols across all files
 *     "package_name":  "axios",
 *     "language":      "javascript",
 *     "warnings":      [...],
 *     "verdict":       "zero_impact" | "has_impact" | "scan_errored",
 *     "verdict_reason":"..."
 *   }
 *
 * The `verdict` field lets the skill (Phase 4) short-circuit without
 * re-deriving conclusions from the raw array. Schema first introduced
 * in ast_scanner_go.go.
 *
 * Import patterns covered:
 *   ESM:  import X from 'pkg'                  -> esm_default
 *         import { Y as Z } from 'pkg'         -> esm_named (records original Y)
 *         import * as ns from 'pkg'            -> esm_namespace
 *         import 'pkg'                         -> esm_side_effect
 *         import type { T } from 'pkg'         -> esm_type_only (flagged)
 *   CJS:  const X = require('pkg')             -> cjs_default
 *         const { Y } = require('pkg')         -> cjs_destructure
 *   Dyn:  await import('pkg')                  -> dynamic
 *
 * Submodule imports ('pkg/foo/bar') are also matched and recorded with the
 * full submodule path so Phase 3 can distinguish "deep import" usage.
 */

'use strict';

const fs = require('fs');
const path = require('path');

let parser, traverse;
try {
    parser = require('@babel/parser');
    traverse = require('@babel/traverse').default;
} catch (err) {
    process.stderr.write(
        'ERROR: @babel/parser / @babel/traverse not installed.\n' +
        'Run: cd <skill-dir>/scripts && npm install\n' +
        `(${err.message})\n`
    );
    process.exit(2);
}

const SKIP_DIRS = new Set([
    'node_modules', '.git', 'dist', 'build', 'out', '.next', '.nuxt',
    '.cache', 'coverage', '.turbo', '.parcel-cache',
]);
const SOURCE_EXTS = new Set(['.js', '.jsx', '.mjs', '.cjs', '.ts', '.tsx']);

function isTargetModule(modulePath, packageName) {
    if (!modulePath || typeof modulePath !== 'string') return false;
    if (modulePath === packageName) return true;
    if (modulePath.startsWith(`${packageName}/`)) return true;
    return false;
}

function getContext(sourceLines, lineno, radius = 5) {
    const start = Math.max(0, lineno - radius - 1);
    const end = Math.min(sourceLines.length, lineno + radius);
    const lines = sourceLines.slice(start, end);
    return lines.map((line, i) =>
        `${String(start + i + 1).padStart(4, ' ')} | ${line}`
    ).join('\n');
}

function parseFile(filepath, source) {
    const ext = path.extname(filepath);
    const isTS = ext === '.ts' || ext === '.tsx';
    const isJSX = ext === '.jsx' || ext === '.tsx';
    const plugins = [];
    if (isTS) plugins.push('typescript');
    if (isJSX) plugins.push('jsx');
    // Permissive plugin set that covers most real-world code.
    plugins.push('decorators-legacy', 'classProperties', 'topLevelAwait', 'importAttributes');
    try {
        return parser.parse(source, {
            sourceType: 'unambiguous',
            allowReturnOutsideFunction: true,
            allowImportExportEverywhere: true,
            allowAwaitOutsideFunction: true,
            errorRecovery: true,
            plugins,
        });
    } catch (err) {
        // Fallback: try without TS/JSX plugins if a syntax we don't understand trips parsing
        try {
            return parser.parse(source, {
                sourceType: 'unambiguous',
                errorRecovery: true,
                plugins: ['jsx'],
            });
        } catch (_) {
            return null;
        }
    }
}

// Returns { result, parseFailed }. result is null when the file parsed but
// contained no match; parseFailed is true when the file was unreadable or
// the parser bailed even after fallback. Caller needs the distinction to
// pick between `zero_impact` and `scan_errored` verdicts.
function scanFileInternal(filepath, packageName) {
    let source;
    try {
        source = fs.readFileSync(filepath, 'utf8');
    } catch (_) {
        return { result: null, parseFailed: true };
    }
    const ast = parseFile(filepath, source);
    if (!ast) return { result: null, parseFailed: true };
    const sourceLines = source.split('\n');

    const imports = [];
    const usages = [];
    // localName -> { sourceModule, originalSymbol, importType }
    const importedNames = new Map();
    let hasMatch = false;

    function recordImport(entry, localName, info) {
        imports.push(entry);
        if (localName && info) {
            importedNames.set(localName, info);
        }
        hasMatch = true;
    }

    traverse(ast, {
        // import X from 'pkg' | import { Y } from 'pkg' | import * as ns from 'pkg' | import 'pkg'
        ImportDeclaration(p) {
            const node = p.node;
            const source = node.source.value;
            if (!isTargetModule(source, packageName)) return;
            const lineno = node.loc ? node.loc.start.line : 0;
            const ctx = getContext(sourceLines, lineno);

            if (node.specifiers.length === 0) {
                recordImport({
                    type: 'esm_side_effect',
                    module: source,
                    line: lineno,
                    context: ctx,
                }, null, null);
                return;
            }

            for (const spec of node.specifiers) {
                const isTypeOnly = node.importKind === 'type' || spec.importKind === 'type';
                if (spec.type === 'ImportDefaultSpecifier') {
                    recordImport({
                        type: isTypeOnly ? 'esm_type_only' : 'esm_default',
                        module: source,
                        local: spec.local.name,
                        line: lineno,
                        context: ctx,
                    }, spec.local.name, { sourceModule: source, originalSymbol: 'default', importType: 'esm_default' });
                } else if (spec.type === 'ImportNamespaceSpecifier') {
                    recordImport({
                        type: isTypeOnly ? 'esm_type_only' : 'esm_namespace',
                        module: source,
                        local: spec.local.name,
                        line: lineno,
                        context: ctx,
                    }, spec.local.name, { sourceModule: source, originalSymbol: '*', importType: 'esm_namespace' });
                } else if (spec.type === 'ImportSpecifier') {
                    const original = spec.imported.type === 'Identifier'
                        ? spec.imported.name
                        : spec.imported.value;
                    recordImport({
                        type: isTypeOnly ? 'esm_type_only' : 'esm_named',
                        module: source,
                        imported: original,
                        local: spec.local.name,
                        line: lineno,
                        context: ctx,
                    }, spec.local.name, { sourceModule: source, originalSymbol: original, importType: 'esm_named' });
                }
            }
        },

        // const X = require('pkg') | const { Y } = require('pkg')
        VariableDeclarator(p) {
            const node = p.node;
            if (!node.init) return;
            // Match require('pkg') and require('pkg/sub')
            if (node.init.type !== 'CallExpression') return;
            const callee = node.init.callee;
            if (!callee || callee.name !== 'require') return;
            const arg = node.init.arguments[0];
            if (!arg || arg.type !== 'StringLiteral') return;
            if (!isTargetModule(arg.value, packageName)) return;
            const lineno = node.loc ? node.loc.start.line : 0;
            const ctx = getContext(sourceLines, lineno);

            if (node.id.type === 'Identifier') {
                recordImport({
                    type: 'cjs_default',
                    module: arg.value,
                    local: node.id.name,
                    line: lineno,
                    context: ctx,
                }, node.id.name, { sourceModule: arg.value, originalSymbol: 'module.exports', importType: 'cjs_default' });
            } else if (node.id.type === 'ObjectPattern') {
                for (const prop of node.id.properties) {
                    if (prop.type !== 'ObjectProperty') continue;
                    const keyName = prop.key.type === 'Identifier' ? prop.key.name : prop.key.value;
                    const localName = prop.value.type === 'Identifier' ? prop.value.name : keyName;
                    recordImport({
                        type: 'cjs_destructure',
                        module: arg.value,
                        imported: keyName,
                        local: localName,
                        line: lineno,
                        context: ctx,
                    }, localName, { sourceModule: arg.value, originalSymbol: keyName, importType: 'cjs_destructure' });
                }
            }
        },

        // import('pkg') dynamic
        CallExpression(p) {
            const node = p.node;
            if (node.callee.type === 'Import') {
                const arg = node.arguments[0];
                if (arg && arg.type === 'StringLiteral' && isTargetModule(arg.value, packageName)) {
                    const lineno = node.loc ? node.loc.start.line : 0;
                    imports.push({
                        type: 'dynamic',
                        module: arg.value,
                        line: lineno,
                        context: getContext(sourceLines, lineno),
                    });
                    hasMatch = true;
                }
            }
            // require.resolve('pkg')
            if (node.callee.type === 'MemberExpression' &&
                node.callee.object.name === 'require' &&
                node.callee.property.name === 'resolve') {
                const arg = node.arguments[0];
                if (arg && arg.type === 'StringLiteral' && isTargetModule(arg.value, packageName)) {
                    const lineno = node.loc ? node.loc.start.line : 0;
                    imports.push({
                        type: 'cjs_resolve',
                        module: arg.value,
                        line: lineno,
                        context: getContext(sourceLines, lineno),
                    });
                    hasMatch = true;
                }
            }
        },
    });

    // Second pass: track usage sites for every localName we recorded.
    if (importedNames.size > 0) {
        traverse(ast, {
            MemberExpression(p) {
                const node = p.node;
                // Build chain: a.b.c.d -> "a.b.c.d"
                const chain = [];
                let cur = node;
                while (cur && cur.type === 'MemberExpression') {
                    if (cur.property.type === 'Identifier') chain.unshift(cur.property.name);
                    else if (cur.property.type === 'StringLiteral') chain.unshift(cur.property.value);
                    cur = cur.object;
                }
                if (!cur || cur.type !== 'Identifier') return;
                const root = cur.name;
                const info = importedNames.get(root);
                if (!info) return;
                const lineno = node.loc ? node.loc.start.line : 0;
                // Symbol: <pkg>.<originalSymbol>.<chain...> when default/namespace;
                // for named imports the chain hangs directly off the original symbol.
                let symbol;
                if (info.importType === 'esm_namespace') {
                    symbol = `${info.sourceModule}.${chain.join('.')}`;
                } else if (info.originalSymbol === 'default' || info.originalSymbol === 'module.exports') {
                    symbol = `${info.sourceModule}.default${chain.length ? '.' + chain.join('.') : ''}`;
                } else {
                    symbol = `${info.sourceModule}.${info.originalSymbol}${chain.length ? '.' + chain.join('.') : ''}`;
                }
                usages.push({
                    symbol,
                    line: lineno,
                    context: getContext(sourceLines, lineno),
                });
            },
            Identifier(p) {
                // Only count "referenced" identifiers: usages, not binding
                // sites. @babel/traverse provides isReferencedIdentifier()
                // which already filters out import specifiers, declaration
                // patterns, object keys, etc.
                if (!p.isReferencedIdentifier()) return;
                // MemberExpression roots are handled by the MemberExpression
                // visitor above — skip to avoid duplicates.
                if (p.parent && p.parent.type === 'MemberExpression' && p.parent.object === p.node) return;
                const node = p.node;
                const info = importedNames.get(node.name);
                if (!info) return;
                const lineno = node.loc ? node.loc.start.line : 0;
                let symbol;
                if (info.originalSymbol === 'default' || info.originalSymbol === 'module.exports') {
                    symbol = `${info.sourceModule}.default`;
                } else if (info.originalSymbol === '*') {
                    symbol = `${info.sourceModule}.*`;
                } else {
                    symbol = `${info.sourceModule}.${info.originalSymbol}`;
                }
                usages.push({
                    symbol,
                    line: lineno,
                    context: getContext(sourceLines, lineno),
                });
            },
        });
    }

    if (!hasMatch) return { result: null, parseFailed: false };
    return { result: { file: filepath, imports, usages }, parseFailed: false };
}

// Walks the project, mutating `stats` { results[], filesScanned, parseErrors }.
function walkProject(projectPath, packageName, stats) {
    let entries;
    try {
        entries = fs.readdirSync(projectPath, { withFileTypes: true });
    } catch (_) {
        return;
    }
    for (const entry of entries) {
        if (entry.name.startsWith('.')) continue;
        const full = path.join(projectPath, entry.name);
        if (entry.isDirectory()) {
            if (SKIP_DIRS.has(entry.name)) continue;
            walkProject(full, packageName, stats);
        } else if (entry.isFile()) {
            const ext = path.extname(entry.name);
            // Skip .d.ts in the project — those are type re-exports, not usage
            if (entry.name.endsWith('.d.ts')) continue;
            if (!SOURCE_EXTS.has(ext)) continue;
            stats.filesScanned += 1;
            const { result, parseFailed } = scanFileInternal(full, packageName);
            if (parseFailed) {
                stats.parseErrors += 1;
            } else if (result) {
                stats.results.push(result);
            }
        }
    }
}

function computeVerdict(results, filesScanned, parseErrors, packageName) {
    const importCount = results.reduce((s, r) => s + r.imports.length, 0);
    const usageCount = results.reduce((s, r) => s + r.usages.length, 0);

    if (importCount === 0 && usageCount === 0 && parseErrors === 0) {
        return {
            verdict: 'zero_impact',
            verdict_reason: `scanned ${filesScanned} source file(s); no import or usage of ${packageName} found`,
            import_count: importCount,
            usage_count: usageCount,
        };
    }
    if (importCount === 0 && usageCount === 0 && parseErrors > 0) {
        return {
            verdict: 'scan_errored',
            verdict_reason: `${parseErrors} file(s) failed to parse; no matches in the remaining ${filesScanned - parseErrors}. Treat as inconclusive.`,
            import_count: importCount,
            usage_count: usageCount,
        };
    }
    return {
        verdict: 'has_impact',
        verdict_reason: `${importCount} import site(s) and ${usageCount} usage(s) found across ${results.length} file(s)`,
        import_count: importCount,
        usage_count: usageCount,
    };
}

function main() {
    const [, , projectPath, packageName] = process.argv;
    if (!projectPath || !packageName) {
        process.stderr.write('Usage: node ast_scanner_js.js <project_path> <package_name>\n');
        process.exit(1);
    }
    const stats = { results: [], filesScanned: 0, parseErrors: 0 };
    walkProject(projectPath, packageName, stats);

    const { verdict, verdict_reason, import_count, usage_count } =
        computeVerdict(stats.results, stats.filesScanned, stats.parseErrors, packageName);

    const warnings = [];
    if (stats.parseErrors > 0) {
        warnings.push(`${stats.parseErrors} file(s) failed to parse; results may be incomplete`);
    }

    process.stdout.write(JSON.stringify({
        scan_results: stats.results,
        total_files: stats.filesScanned,
        files_scanned: stats.filesScanned,
        import_count,
        usage_count,
        package_name: packageName,
        language: 'javascript',
        warnings,
        verdict,
        verdict_reason,
    }, null, 2) + '\n');
}

main();
